docs/specs/app2app.md

```diff
@@ -0,0 +1,199 @@
+    Title Android App2App Flow
+    participant B as App B
+    participant O as Android
+    participant A as App A
+    participant S as Authgear Server
+
+
+    rect rgb(191, 223, 255)
+    note right of B: authgear.startApp2AppAuthentication(authorizeUri: String, redirectUri: String)
+    B ->> O: Intent (authorizeUri)
+    end
+
+    O ->> A: Applink (authorizeUri)
+
+    rect rgb(191, 223, 255)
+    note right of A: authgear.handleApp2AppAuthenticationRequest(uri: Uri)
+    A ->> S: Request new authorization code
+    S ->> A: Authorization code
+    A ->> O: Intent (redirectUri + authCode)
+    end
+    O ->> B: Applink (redirectUri + authCode)
+
+    rect rgb(191, 223, 255)
+    note right of B: authgear.handleApp2AppAuthenticationResult(uri: Uri)
+    B ->> S: Exchange token
+    S ->> B: Refresh token & Access token
+    end
+```
+
+In android, app2app works with the following flow:
+
+Assume there are two apps, A and B. App A is holding a valid user session, and app B is trying to authenticate through app A.
+
+- App A
+
+  - The client id is `client_a`
+  - App A is able to handle applinks in domain `https://a.example.com/` with a properly setup [.well-known/assetlinks.json](https://developer.android.com/training/app-links/verify-android-applinks) file.
+
+- App B
+
+  - The configured `x_redirect_uris` is `["appb://redirect"]`
+  - The package name is `com.example.appb`
+  - The client id is `client_b`
+  - App A is able to handle applinks in domain `https://b.example.com/` with a properly setup [.well-known/assetlinks.json](https://developer.android.com/training/app-links/verify-android-applinks) file.
+
+1. App B initiates an app2app authentication flow using an intent with uri `https://a.example.com/authorize?client_id=client_a&redirect_uri=https%3A%2F%2Fb.example.com%2Fredirect&code_challenge=CHALLENGE`.
+
+   - The parameters of the uri are:
+
+     - `client_id`: The client id of the initiating app.
+     - `redirect_uri`: The redirect uri used to recieve the authentication result.
+     - `code_challenge`: The PKCE code challenge with the same purpose in the oidc authorize endpoint.
+
+   - For security reason, App B should verify the signature of the app that is going to handle the intent by:
+     - Fetching the `.well-known/assetlinks.json` from the domain of the uri. In this example it will be `https://a.example.com/.well-known/assetlinks.json`.
+     - Obtain the package name and signature of the app which is going to handle the uri.
+     - Find the `"delegate_permission/common.handle_all_urls"` relation, and obtain the related `package_name` and `sha256_cert_fingerprints` from the section. Verify it with the package name and signature obtained in the previous step. If none of the items match, returns error.
+     - After validated the package name and signature, start an activity with an intent with the above uri.
+
+2. App A handles the intent.
+
+   - Returns an error if App A doesn't have an valid session.
```

Comment:
I think we should support first `authenticate()` in App A, after that, continue the flow.

Reply:
@louischan-oursky We don't know what is the parameter we should use to call `authenticate()` in AppA in app2app flow. If we want to support that, we may need to support all parameters of `authenticate()` in `startApp2AppAuthentication()`.

Reply:
Ideally, we should let the developer to decide how to call `authenticate()`. What we need to do is to resume the app2app2 flow after authentication.

Reply:
@louischan-oursky Then I think maybe we don't need `startApp2AppAuthentication`, we can just reuse `authenticate` with an endpoint which will be handled by App A.

Reply:
This flow can actually handled by the current api design by calling `authenticate` in app A before calling `handleApp2AppAuthenticationRequest`, will write it down in the spec.